### PR TITLE
[lldb][DWARF] Do not complete type from declaration die.

### PR DESCRIPTION
**Pull request by ZequanWu**
Fix the problem: https://github.com/llvm/llvm-project/pull/90663#issuecomment-2105164917 by enhancing a double-check for #90663

### CODE REVIEW

**Review by clayborg — COMMENTED** [review]
Let me verify this works. I would also like this to fix:
```
bool DebugNamesDWARFIndex::ProcessEntry(
    const DebugNames::Entry &entry,
    llvm::function_ref<bool(DWARFDIE die)> callback) {
  std::optional<DIERef> ref = ToDIERef(entry);
  if (!ref)
    return true;
  SymbolFileDWARF &dwarf = *llvm::cast<SymbolFileDWARF>(
      m_module.GetSymbolFile()->GetBackingSymbolFile());
  DWARFDIE die = dwarf.GetDIE(*ref);
  if (!die)
    return true;
  // Watch out for forward declarations that appear in the .debug_names tables
  // only due to being there for a DW_IDX_parent.
  if (die.GetAttributeValueAsUnsigned(DW_AT_declaration, 0))
    return true;
  return callback(die);
}
```
This adds:
```
  // Watch out for forward declarations that appear in the .debug_names tables
  // only due to being there for a DW_IDX_parent.
  if (die.GetAttributeValueAsUnsigned(DW_AT_declaration, 0))
    return true;
```
To the above function to ensure we don't waste any time trying to parse any DIE information for forward declaration when using .debug_names. This will cause a TON of churn on our DWARF parser and cause us to pull in and parse way too much.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp [hunk 1]
```diff
@@ -2343,7 +2343,10 @@ bool DWARFASTParserClang::CompleteTypeFromDWARF(const DWARFDIE &die,
   // clang::ExternalASTSource queries for this type.
   m_ast.SetHasExternalStorage(clang_type.GetOpaqueQualType(), false);
 
-  if (!die)
+  ParsedDWARFTypeAttributes attrs(die);
+  bool is_forward_declaration = IsForwardDeclaration(
+      die, attrs, SymbolFileDWARF::GetLanguage(*die.GetCU()));
+  if (!die || is_forward_declaration)
```
Comment (clayborg) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:2349]: Maybe this should first check if `die` is valid, then do this?:
```
  if (!die)
    return false;
  ParsedDWARFTypeAttributes attrs(die);
  if (IsForwardDeclaration(die, attrs, SymbolFileDWARF::GetLanguage(*die.GetCU())))
    return false;
```

**Comment by ZequanWu** [comment]
> To the above function to ensure we don't waste any time trying to parse any DIE information for forward declaration when using .debug_names. This will cause a TON of churn on our DWARF parser and cause us to pull in and parse way too much.

That sounds like a better fix.

**Review by clayborg — APPROVED** [review]
This does fix things on your side. I will take care of a new PR for not searching all definition DIEs from the .debug_names.

**Pull request merged**